docs: add resource types reference page (#25177)
kubernetes/website: content/en/docs/reference/resource-types/_index.md

---
title: Resource Types
content_type: reference
---

<!-- overview -->

This page provides information on the different resource types you can request, built into Kubernetes.

<!-- body -->

There are several different resource types: `cpu`, `memory`, `storage`, `hugepages-*`, and `emphemeral-storage`.

Resource requests on Pods are an integral part of scheduling and 
limiting and managing resource consumption in a Kubernetes 
cluster. 
See [Kubernetes Scheduler](/docs/concepts/scheduling-eviction/kube-scheduler/) and [Requests and Limits](/docs/concepts/configuration/manage-resources-containers/#requests-and-limits) for more details.

> ## CPU 

Reviewer: nit: Would you add a newline before the CPU heading? Thanks!

Author: Done!

> ## CPU 

Author: ```suggestion

## CPU 
```

---

> title: Resource Types

Reviewer: I think the title of the page (and the directory) should be changed. I was initially confused
by the use of the "resource".

Reviewer 2: Would "PersistentVolumeClaim Resource Types" clarify it further?

content_type: reference
---

<!-- overview -->

This page provides information on the different resource types you can request, built into Kubernetes.

> <!-- body -->

Reviewer 3: This is a short overview IMO. Best to merge it with the paragraphs that follow. That way “CPU” is the first part of the page body.

There are several different resource types: `cpu`, `memory`, `storage`, `hugepages-*`, and `emphemeral-storage`.

Resource requests on Pods are an integral part of scheduling and 
limiting and managing resource consumption in a Kubernetes 
cluster. 
See [Kubernetes Scheduler](/docs/concepts/scheduling-eviction/kube-scheduler/) and [Requests and Limits](/docs/concepts/configuration/manage-resources-containers/#requests-and-limits) for more details.

> ## CPU 

Reviewer 3: How about making these be a 3rd level heading, under a 2nd level heading “resource types”.

In contrast with cpu, when the memory limit is reached, the 
Pod is killed.

See [Meaning of Memory](/docs/concepts/configuration/manage-resources-containers/#meaning-of-memory) for more details on what memory means in kubernetes.

## Storage 
The `storage` resource represents volume size in bytes and is specified with a number and a binary suffix. e.g. `500Gi` = `500GiB` = `500 * 1024 * 1024 * 1024`

You can use the `storage` resource in a [PersistentVolumeClaim](/docs/concepts/storage/persistent-volumes/#introduction).

## Huge pages

On Kubernetes v1.14 and newer you can request the `hugepages-*` 
resource. Huge pages are a Linux-specific feature where the node 
kernel allocates blocks of memory that are much larger than the 
default page size.

You cannot overcommit `hugepages-*` resources. This is different from the `memory` and `cpu` resources.

You request huge pages with a combination of two binary
suffixes. The first is part of the resource name in the request: 
e.g. `hugepages-2Mi`. It specifies the size of the pages. The second part is the value of allocatable pages e.g. `80Mi`. 

> ## Emphemeral storage

Reviewer 3: The correct spelling is “Ephemeral”.

Emphemeral storage is storage provided to pods that has no 

> long-term gurantee of durablity. Empheral Storage is in beta. 

Reviewer 3: ```suggestion
long-term guarantee of durability. Ephemeral storage is in beta.
```

Empheral storage is represented in bytes and is specified with a 
number and a binary suffix. e.g. `500Mi` = `500MiB` = `500 * 1024 * 1024`. 

You can specify the `ephemeral-storage` resource in Pods and any object that embeds a Pod template.

See [Local emphemeral storage](/docs/concepts/configuration/manage-resources-containers/#local-ephemeral-storage) for more details.

> 

Reviewer 3: We should add a heading for custom (extended) resource types - ie, https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#extended-resources

---

> title: Resource Types

Reviewer 3: ```suggestion
title: Kubernetes Resource Management Types
```
?

content_type: reference
---

<!-- overview -->

This page provides information on the different resource types you can request, built into Kubernetes.

There are several different resource types: `cpu`, `memory`, `storage`, `hugepages-*`, and `ephemeral-storage`.

Resource requests on Pods are an integral part of scheduling and 
limiting and managing resource consumption in a Kubernetes 

> cluster. 

Author: I realize it's a duplication but the value it adds imo is the locality to the references and examples. The original page is an in-depth explanation of the resources

Happy to here an alternative. Should I just remove the overview or reword it?

See [Kubernetes Scheduler](/docs/concepts/scheduling-eviction/kube-scheduler/) and [Requests and Limits](/docs/concepts/configuration/manage-resources-containers/#requests-and-limits) for more details.

<!-- body -->

## CPU 
The `cpu` resource represents CPU in cores and can be specified as either a decimal (e.g. `1`, `0.5`) or as millicores (e.g. `500m`, `1000m`). `0.5` = `500m`.

You can request or limit the `cpu` resource for Pods and their containers, and any for object that embeds a [Pod template](/docs/concepts/workloads/pods/#pod-templates).

> When the cpu limit is reached, the Pod is throttled.
> 
> See [Meaning of CPU](/docs/concepts/configuration/manage-resources-containers/#meaning-of-cpu) for more details on what CPU means in kubernetes.

Reviewer 3: (nit)
```suggestion
The container runtime ensures that processes in a container do not exceed their CPU allocation.

See [Meaning of CPU](/docs/concepts/configuration/manage-resources-containers/#meaning-of-cpu)
for details of CPU units in Kubernetes.
```

## Memory
The `memory` resource represents memory in bytes and is specified with a number and a binary suffix. e.g. `500Mi` = `500MiB` = `500 * 1024 * 1024`. 

You can request or limit the `memory` resource in Pods and their containers, and any object that embeds a Pod template.

> In contrast with cpu, when the memory limit is reached, the 
> Pod is killed.

Reviewer 3: How about:
```suggestion
If a container or Pod tries to use more memory than the configured limit, the
container runtime and / or host operating system kernel typically intervenes by
stopping a process in that Pod. That mechanism is called
“out of memory kill”, and is different from how CPU resource limits are handled

(CPU limits are enforced using time-slicing, whereas memory limits are enforced
by stopping containers that try to violate them).
```

> See [Meaning of Memory](/docs/concepts/configuration/manage-resources-containers/#meaning-of-memory) for more details on what memory means in kubernetes.

Reviewer 3: nit:
```suggestion
See [Meaning of Memory](/docs/concepts/configuration/manage-resources-containers/#meaning-of-memory)
for more details on what memory means in Kubernetes.
```

## Storage 
The `storage` resource represents volume size in bytes and is specified with a number and a binary suffix. e.g. `500Gi` = `500GiB` = `500 * 1024 * 1024 * 1024`

> You can use the `storage` resource in a [PersistentVolumeClaim](/docs/concepts/storage/persistent-volumes/#introduction).

Reviewer 3: (nit)
```suggestion
You can use the `storage` resource in a [PersistentVolumeClaim](/docs/concepts/storage/persistent-volumes/#persistentvolumeclaims).
```
or, even better:
```suggestion
You can use the `storage` resource in a
{{< glossary_tooltip text="PersistentVolumeClaim" term_id="persistent-volume-claim" >}}.
```

> You can use the `storage` resource in a [PersistentVolumeClaim](/docs/concepts/storage/persistent-volumes/#introduction).

Reviewer 4: The verb 'use' is misleading.

## Huge pages

> On Kubernetes v1.14 and newer you can request the `hugepages-*` 
> resource. Huge pages are a Linux-specific feature where the node 
> kernel allocates blocks of memory that are much larger than the 
> default page size.

Reviewer 3: Can we assume the reader is using a supported version of Kubernetes?
```suggestion
Huge pages are a Linux-specific feature where the node 
kernel allocates blocks of memory that are much larger than the 
default page size. You can set requests or limits for particular sizes
of huge page.
```

Author: I think that's fair. I just like to thing of a newcomer that may be following old tutorials and they fall into some unclear errors

You cannot overcommit `hugepages-*` resources. This is different from the `memory` and `cpu` resources.

> You request huge pages with a combination of two binary

Reviewer 4: Referring this as 'a combination' is confusing.
The paragraphs above are an exact copy of https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#resource-types.

suffixes. The first is part of the resource name in the request: 

> e.g. `hugepages-2Mi`. It specifies the size of the pages. The second part is the value of allocatable pages e.g. `80Mi`. 

Reviewer 3: for example, `hugepages-2Mi`.
That `2Mi` specifies the size of the pages; in this case, 2 MiB.
The second part is the value of allocatable pages. For example, `hugepages-2Mi: 100Mi`
allows using up to 50 huge pages, 2 MiB each, for a total of 100 MiB.
See the [example](#example-hugepages) to see this in context.

## Ephemeral storage

Ephemeral storage is storage provided to pods that has no 
long-term guarantee of durability. Empheral Storage is in beta.
Empheral storage is represented in bytes and is specified with a 
number and a binary suffix. e.g. `500Mi` = `500MiB` = `500 * 1024 * 1024`. 

You can specify the `ephemeral-storage` resource in Pods and any object that embeds a Pod template.

See [Local ephemeral storage](/docs/concepts/configuration/manage-resources-containers/#local-ephemeral-storage) for more details.

## Extended Resources

Extended resources are fully-qualified resource names outside the kubernetes.io domain. 
They allow cluster operators to advertise and users to consume the non-Kubernetes-built-in resources.

See [Extended Resources](/docs/concepts/configuration/manage-resources-containers/#extended-resources) for more information.

## Examples

> ### CPU and memory

Reviewer 3: ```suggestion
### CPU and memory {#example-cpu-memory}
```

```yaml
apiVersion: v1
kind: Pod
metadata:
  ...
spec:
  ...
  requests:
    cpu: 500m
    memory: 250Mi
  limits:
    cpu: 700m
    memory: 500Mi
```

> ### Storage

Reviewer 3: ```suggestion
### Storage {#example-storage}
```

```yaml
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  ...
spec:
  ...
  requests:
    storage: 20Gi
```

> ### Huge pages

Reviewer 3: ```suggestion
### Huge pages {#example-hugepages}
```

```yaml
apiVersion: v1
kind: Pod
metadata:
  ...
spec:
  ...
  requests:
    hugepages-2Mi: 80Mi
  limits:
    hugepages-2Mi: 100Mi
```

> ### Ephemeral Storage

Reviewer 3: ```suggestion
### Ephemeral storage {#example-ephemeral-storage}
```

> ---
> title: Resource Types
> ---

Reviewer 3: ```suggestion
---
title: Resource Types
content_type: reference
---
```

Author: Where can I find the list of `content_type`s?

<!-- overview -->

> This section provides information on the different resource types you can request, built into Kubernetes.

Reviewer 3: ```suggestion
This page provides information on the different resource types you can request, built into Kubernetes.
```

Reviewer 3: Maybe this page should also mention that [extended resources](https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#extended-resources) exist?

Author: I agree!

<!-- body -->

> There are 3 different resource types: `cpu`, `memory`, and `storage`.

Reviewer 3: Hi. There is also `hugepages-*` - see https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#resource-requests-and-limits-of-pod-and-container

and, as a beta feature, `ephemeral-storage` - see https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#local-ephemeral-storage and https://kubernetes.io/docs/concepts/policy/resource-quotas/#storage-resource-quota

Author: Copy, I'll add these!

> 

Reviewer 3: I recommend telling the reader that resources are relevant to [scheduling](https://kubernetes.io/docs/concepts/scheduling-eviction/) as well as for enforcing limits.

Author: Sounds good!

## CPU 

> The `cpu` resource represents CPU in cores and can be specified as either a decimal (e.g. `1`, `0.5`) or as millicores (e.g. `500m`, `1000m`). `0.5` = `500m`.

Reviewer 3: I recommend explaining “CPU” in more detail; see https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#meaning-of-cpu for some wording you could migrate to this page.

Author: I believe this was resolved below. Feel free to re-add this comment if you feel it wasn't addressed!

> You can use the cpu resource in Pods and any object that embeds a Pod template.

Reviewer 3: ```suggestion
You can request or limit the `cpu` resource for Pods and their containers, and any for object that embeds a [Pod template](https://kubernetes.io/docs/concepts/workloads/pods/#pod-templates).
```

## Memory
The `memory` resource represents memory in bytes and is specified with a number and a binary suffix. e.g. `500Mi` = `500MiB` = `500 * 1024 * 1024`. 

> You can use the memory resource in Pods and any object that embeds a Pod template.

Reviewer 3: ```suggestion
You can request or limit the `memory` resource for Pods and their containers, and any for object that embeds a Pod template.
```

## Storage 
The `storage` resource represents volume size in bytes and is specified with a number and a binary suffix. e.g. `500Gi` = `500GiB` = `500 * 1024 * 1024 * 1024`

> You can use the storage resource in PersistentVolumeClaims.

Reviewer 3: ```suggestion
You can use the `storage` resource in a PersistentVolumeClaim.
```

Examples: 

> ### CPU and Memory

Reviewer 3: nit:
```suggestion
### CPU and memory
```
as [headings](https://kubernetes.io/docs/contribute/style/style-guide/#headings) are usually sentence case.

## CPU 
The `cpu` resource represents CPU in cores and can be specified as either a decimal (e.g. `1`, `0.5`) or as millicores (e.g. `500m`, `1000m`). `0.5` = `500m`.

> You can request or limit the `cpu` resource for Pods and their containers, and any for object that embeds a [Pod template](https://kubernetes.io/docs/concepts/workloads/pods/#pod-templates).

Reviewer 3: We always use relative hyperlinks:
```suggestion
You can request or limit the `cpu` resource for Pods and their containers, and any for object that embeds a [Pod template](/docs/concepts/workloads/pods/#pod-templates).
```
